### PR TITLE
add a proto `nupm publish` command to generate list of files

### PR DESCRIPTION
**Pull request by amtoine**
followup to
- https://github.com/nushell/nupm/pull/4

## description

this PR implements a part of the `nupm publish` command by following the [metadata file specification](https://github.com/nushell/nupm/blob/main/docs/design/references/METADATA.md#automatically-generated-outside-of-the-user-created-metadata-file).

it implements the automatic generation of the list of package files meant to be used by `nupm install` :+1: 

## example output metadata files files
- running either one of the following in [`amtoine/nu-git-manager`](https://github.com/amtoine/nu-git-manager)
    - `nupm publish --generate-metadata --host github --repo amtoine/nu-git-manager --revision main`
    - `nupm publish --generate-metadata --host github --repo amtoine/nu-git-manager --revision main --path .`
```
╭────────┬─────────────────────────────────────────────────────────────────────┬─────────────────────────────────┬───────────────────────────────────────────────────────────────────────────────────────────────┬────────────────────────╮
│      # │                              checksum                               │              name               │                                            raw-url                                            │      supported-os      │
├────────┼─────────────────────────────────────────────────────────────────────┼─────────────────────────────────┼───────────────────────────────────────────────────────────────────────────────────────────────┼────────────────────────┤
│      0 │ 3972dc9744f6499f0f9b2dbf76696f2ae7ad8af9b23dde66d6af86c9dfb36986    │ LICENSE                         │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/LICENSE                         │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│      1 │ 405eddc303110b2346d5b454798f615f5408456c5dded19a2be97b6b4b93216d    │ README.md                       │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/README.md                       │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│      2 │ b134b6a88857b7904146210d0551a36745300892c3d6be250dcfbb77bfa8ec17    │ docs/README.md                  │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/docs/README.md                  │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│      3 │ 25d06df1022c6c113ef1c12bcbf59595f3b990595f846488d71a3b03b8bfbf45    │ docs/installation/README.md     │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/docs/installation/README.md     │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│      4 │ 91f0d0c38463a4f05018e3b2b897e0ee51c3cd470bffc12a45d6431d987fce19    │ docs/installation/manual.md     │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/docs/installation/manual.md     │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│      5 │ 7844f19b80ab6ad8afb9a1877d5792e51e31e914678d570984f0590c6b0468a0    │ docs/installation/nupm.md       │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/docs/installation/nupm.md       │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│      6 │ fe9433785984f01999654d9edeeedbf895bbf36294f9a021af7968aa180bc012    │ mod.nu                          │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/mod.nu                          │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│      7 │ 317c2b86f0456d94a0b9466a78c48962a04701c3d74b7dea51e6c9eaedbfbbcd    │ src/gm/mod.nu                   │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/src/gm/mod.nu                   │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│      8 │ bf3e20d4ff9722ad0cdd8b938aadadd9de07045ab8b39f1e338dfb371f6051ce    │ src/gm/utils/utils.nu           │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/src/gm/utils/utils.nu           │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│      9 │ 986ea60c315531960f6d8af194364dd4a721c3fac524293f701b08ee45458091    │ src/sugar/completions/git.nu    │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/src/sugar/completions/git.nu    │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│     10 │ e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855    │ src/sugar/completions/mod.nu    │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/src/sugar/completions/mod.nu    │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│     11 │ 94db2551b1ae52e37717ab2c0d138560caed324e45222048bbb2936eb4c9bd17    │ src/sugar/dotfiles.nu           │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/src/sugar/dotfiles.nu           │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│     12 │ 8560ccc994d283be27f94602915218215e03ca4212ba4507ec4d05398b167793    │ src/sugar/gh.nu                 │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/src/sugar/gh.nu                 │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│     13 │ 7c20b26b30e1f012cf4c40f807578e5ebf1f3c039531897913e926e61915578c    │ src/sugar/gist.nu               │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/src/sugar/gist.nu               │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│     14 │ 262de405392672ac641cd9f41ef3aa3784706c972b766a5b5ec789dda37cd168    │ src/sugar/git.nu                │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/src/sugar/git.nu                │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│     15 │ 8bafc2da80cb84751e0afd3250329e79e818bf8294105363e6583f15d4680649    │ src/sugar/mod.nu                │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/src/sugar/mod.nu                │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
│     16 │ 82c599c5fc24317500df037ccd24e8934dd8f8f7d94b1da0e31e5b380567500e    │ toolkit/mod.nu                  │ https://raw.githubusercontent.com/amtoine/nu-git-manager/main/toolkit/mod.nu                  │ ╭────────┬────────╮    │
│        │                                                                     │                                 │                                                                                               │ │ name   │ linux  │    │
│        │                                                                     │                                 │                                                                                               │ │ arch   │ x86_64 │    │
│        │                                                                     │                                 │                                                                                               │ │ family │ unix   │    │
│        │                                                                     │                                 │                                                                                               │ ╰────────┴────────╯    │
╰────────┴─────────────────────────────────────────────────────────────────────┴─────────────────────────────────┴───────────────────────────────────────────────────────────────────────────────────────────────┴────────────────────────╯
```
- running `nupm publish --generate-metadata --host github --repo amtoine/zellij-layouts --revision main --path nu-zellij/` in [`amtoine/zellij-layouts`](https://github.com/amtoine/zellij-layouts)
```
╭────────────┬───────────────────────────────────────────────────────────────────────────┬────────────────────┬────────────────────────────────────────────────────────────────────────────────────────────┬──────────────────────────────╮
│          # │                                 checksum                                  │        name        │                                          raw-url                                           │         supported-os         │
├────────────┼───────────────────────────────────────────────────────────────────────────┼────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────────────┤
│          0 │ 3972dc9744f6499f0f9b2dbf76696f2ae7ad8af9b23dde66d6af86c9dfb36986          │ LICENSE            │ https://raw.githubusercontent.com/amtoine/zellij-layouts/main/nu-zellij/LICENSE            │ ╭────────┬────────╮          │
│            │                                                                           │                    │                                                                                            │ │ name   │ linux  │          │
│            │                                                                           │                    │                                                                                            │ │ arch   │ x86_64 │          │
│            │                                                                           │                    │                                                                                            │ │ family │ unix   │          │
│            │                                                                           │                    │                                                                                            │ ╰────────┴────────╯          │
│          1 │ d6886fc5dea359ee5a72dcb927c5a83b72c32feec1b71d80e0b61e2776e08e4a          │ README.md          │ https://raw.githubusercontent.com/amtoine/zellij-layouts/main/nu-zellij/README.md          │ ╭────────┬────────╮          │
│            │                                                                           │                    │                                                                                            │ │ name   │ linux  │          │
│            │                                                                           │                    │                                                                                            │ │ arch   │ x86_64 │          │
│            │                                                                           │                    │                                                                                            │ │ family │ unix   │          │
│            │                                                                           │                    │                                                                                            │ ╰────────┴────────╯          │
│          2 │ ef8641c3906125ce415c4e33297e1814a62cc717d6fd63397a88d18caed08d97          │ mod.nu             │ https://raw.githubusercontent.com/amtoine/zellij-layouts/main/nu-zellij/mod.nu             │ ╭────────┬────────╮          │
│            │                                                                           │                    │                                                                                            │ │ name   │ linux  │          │
│            │                                                                           │                    │                                                                                            │ │ arch   │ x86_64 │          │
│            │                                                                           │                    │                                                                                            │ │ family │ unix   │          │
│            │                                                                           │                    │                                                                                            │ ╰────────┴────────╯          │
╰────────────┴───────────────────────────────────────────────────────────────────────────┴────────────────────┴────────────────────────────────────────────────────────────────────────────────────────────┴──────────────────────────────╯
```

### CODE REVIEW

**Comment by amtoine** [comment]
cc/ @skelly37 @Yethal

File: nupm/publish.nu [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    --revision: string  # the revision of the repo holding the package (used with `--generate-metadata`)
+] {
+    if $generate_metadata {
+        # TODO: add support for a `.nupmignore` file in the root of a package
+        # to not add files into the package.
+
+        let PACKAGE_FILE = "package.nuon"
+        let METADATA_FILE = "package.files.nuon"
+
+        log info "generating package file metadata file"
```
Comment (skelly37) [inline nupm/publish.nu:44]: "file metadata file"?

File: nupm/publish.nu [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        let package_files = (
+            $package_files | where {|it|
+                let filename = ($it.name | path basename)
+                ($filename != $PACKAGE_FILE) and ($filename != $METADATA_FILE)
+            }
+        )
+
+        log debug "building metadata for package files:"
+        $package_files | each {|file|
+            log debug $file.name
```
Comment (skelly37) [inline nupm/publish.nu:84]:
```suggestion
            log debug $"processing `($file.name)`"
```

File: nupm/publish.nu [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            {
+                checksum: ($file.name | open --raw | hash sha256)
+                name: ($file.name | if $path != "." { str replace $path "" } else {} | str trim --left --char "/")
+                raw-url: ($base_download_url | update path { path join $file.name } | url join)
+                supported-os: ($nu.os-info | reject kernel_version)
```
Comment (skelly37) [inline nupm/publish.nu:89]: Defaulting to current OS is fine but it should be overridable, so I can easily say `it works as long as there's nushell` without playing with the metadatafile manually

File: nupm/publish.nu [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            }
+        }
+        | save --force $METADATA_FILE
+
+        log info $"package file metadata saved in `($METADATA_FILE)`"
+        log warning $"do not forget to commit the `($METADATA_FILE)` metadata file!"
```
Comment (skelly37) [inline nupm/publish.nu:95]:
```suggestion
        log info $"do not forget to commit the `($METADATA_FILE)` metadata file!"
```

Maybe it's just me but `warning` doesn't seem the best fit. I'd just make it info.

Imagine you're filtering the logs and actually use `warning` verbosity. This would be literally the only log from this function call, seems weird.

File: nupm/publish.nu (hunk 1 above)
Comment (amtoine) [reply nupm/publish.nu]: done in 3850c4f :yum:

File: nupm/publish.nu (hunk 2 above)
Comment (amtoine) [reply nupm/publish.nu]: done in 0b841a1 :+1:

File: nupm/publish.nu (hunk 3 above)
Comment (amtoine) [reply nupm/publish.nu:89]: maybe one could just run a command on `package.files.json` for now?

File: nupm/publish.nu (hunk 4 above)
Comment (amtoine) [reply nupm/publish.nu]: i changed it to a `log info` with yellow color in 702294f :innocent:

File: nupm/publish.nu (hunk 3 above)
Comment (skelly37) [reply nupm/publish.nu:89]: For now, I would just add an optional flag to provide valid OSes and default to what you've written here if the flag was not provided

Comment (amtoine) [reply nupm/publish.nu]: should be done in 82aed49 :wink:

File: nupm/publish.nu [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+# > this command is in a very early stage, it is only meant to generate the
+# > metadata of all the files in the package with the `--generate-metadata`
+# > option.
+export def main [
+    --generate-metadata: bool  # only generate the package metadata file
+    --host: string  # where the package is hosted (used with `--generate-metadata`)
+    --path: string = ""  # the path to the package (used with `--generate-metadata`)
+    --repo: string  # the name of the repo holding the package (used with `--generate-metadata`)
+    --revision: string  # the revision of the repo holding the package (used with `--generate-metadata`)
+    --supported-os: list<record<name: string, arch: string, family>>  # the list of all supported OSes
```
Comment (skelly37) [inline nupm/publish.nu:36]:
```suggestion
    --supported-os: list<record<name: string, arch: string, family: string>>  # the list of all supported OSes
```
?

Comment (amtoine) [reply nupm/publish.nu]: 8dbeea4 :eyes:

**Review by skelly37 — APPROVED** [review]
lgtm

**Comment by Yethal** [comment]
lgtm

**Comment by amtoine** [comment]
> Can you split the code into smaller, reusable functions? As it is now it's kinda hard to follow the control flow given you have both code and data mixed together with no clear separation
>
yup i can, could you put that in "requested changes" so that we can see that later when coming back to this PR? :innocent:

**Review by Yethal — CHANGES_REQUESTED** [review]
Can you split the code into smaller, reusable functions? As it is now it's kinda hard to follow the control flow given you have both code and data mixed together with no clear separation

**Comment by amtoine** [comment]
superseded by #86
